contrib/babelfishpg_tsql/src/session.c

```
@@ -177,44 +175,49 @@ set_cur_user_db_and_path(const char *db_name)
+	char		*path;
+	char		*db_name_copy = pstrdup(db_name);
+	char		*schema;
+	char		*physical_schema = NULL;
+	char 		*dbo_schema_name = get_dbo_schema_name(db_name);
 
-	if ((dbo_role_name && strcmp(user, dbo_role_name) == 0))
-	{
-		physical_schema = get_dbo_schema_name(db_name);
-	}
-	else if (guest_role_name && strcmp(user, guest_role_name) == 0)
-	{
-		const char *guest_schema = get_authid_user_ext_schema_name(db_name, "guest");
+	Assert (db_name != NULL && strlen(db_name) > 0);
 
-		if (!guest_schema)
-			guest_schema = "guest";
-		physical_schema = get_physical_schema_name(pstrdup(db_name), guest_schema);
-	}
-	else
+	schema = get_authid_user_ext_schema_name(db_name, user);
+	if (schema)
 	{
-		const char *schema;
-
-		schema = get_authid_user_ext_schema_name(db_name, user);
-		physical_schema = get_physical_schema_name(pstrdup(db_name), schema);
+		physical_schema = get_physical_schema_name(db_name_copy, schema);
+		pfree(schema);
 	}
-
-	path = psprintf(buffer, quote_identifier(physical_schema));
-	SetConfigOption("search_path",
-					path,
-					PGC_SUSET,
-					PGC_S_DATABASE_USER);
+	path = psprintf(PLTSQL_SEARCH_PATH_BUFFER,
+					physical_schema ? quote_identifier(physical_schema) : quote_identifier(dbo_schema_name),
+					quote_identifier(dbo_schema_name));
+
+	if (current_db_search_path)
+		pfree(current_db_search_path);
+	current_db_search_path = MemoryContextStrdup(TopMemoryContext, path);
+
+	/* set search path default so transaction rollback does not affect it */
+	SetConfigOption("search_path", path,
+					PGC_SUSET, PGC_S_SESSION);
+	/*
+	 * Incase of error inside parallel worker we could reach here before transaction
+	 * abort. Do not try to set_config since it will fail inside a parallel worker
+	 * This function is used in code path where error is never expected
+	 */
+	if (!(IsParallelWorker() && IsAbortedTransactionBlockState()))
+		SetConfigOption("search_path", path,
+						PGC_SUSET, PGC_S_DATABASE_USER);
```

1. What is the reason behind keeping GUC source level PGC_S_DATABASE_USER?
2. Can you elaborate a bit more on parallel worker scenario? Will we be able to cover it with testcases?

---

> What is the reason behind keeping GUC source level PGC_S_DATABASE_USER?

USE DB statement is non transactional, it does not rollback on transaction rollback. 
`PGC_S_DATABASE_USER` sets the default value of a GUC for the session, so even if we rolled back an entire transaction which executed a USE DB, we will still end up with the right search path.

This is covered by existing test cases.

---

> Can you elaborate a bit more on parallel worker scenario? Will we be able to cover it with testcases?
Yes, I actually found this issue when I was running some local tests.

CREATE A pltsql function which executed something like SELECT COUNT(*) FROM t WHERE CAST(col AS TINY) < 1000;
Now make a call to this function from another db.

`SELECT * FROM db.dbo.pltsqfunction();`

The entire function will not run inside parallel worker but the SELECT query inside the function will.
To enter the error code path we can add a row in table t which should throw an error during `CAST(col AS TINY)`

---

Added tests